thunder/examine/memory_caculation.py

```diff
@@ -141,6 +145,10 @@ def get_alloc_memory(trc: TraceCtx) -> tuple[int, dict[str, int]]:
     allocated = 0
 
     tensor_to_memory_data = ProxyDict()
+    for arg in tree_iter((trc.args, trc.kwargs)):
```

Comment:
Backward trace holds some tensors in a list on purpose because Python would hold onto the list variable, but not its content. Clearing a list allows `del` calls to free the memory. Here's an example:
```py
import torch

def f(list_of_tensors):
    print(f"0: {torch.cuda.memory_allocated()/1e6} MB")
    x0, x1 = list_of_tensors
    list_of_tensors.clear()
    print(f"1: {torch.cuda.memory_allocated()/1e6} MB")
    z = x0 + 1
    print(f"2: {torch.cuda.memory_allocated()/1e6} MB")
    del x0
    print(f"3: {torch.cuda.memory_allocated()/1e6} MB")
    w = x1 + z + 1
    print(f"4: {torch.cuda.memory_allocated()/1e6} MB")
    del x1, z
    print(f"5: {torch.cuda.memory_allocated()/1e6} MB")
    return w

l = [torch.ones(1000, device="cuda") for _ in range(2)]
out = f(l)
# prints
# 0: 0.008192 MB
# 1: 0.008192 MB
# 2: 0.012288 MB
# 3: 0.008192 MB
# 4: 0.012288 MB
# 5: 0.004096 MB
```
the same doesn't work with tuples because we can't modify tuple content:
```py
import torch

def f(tuple_of_tensors):
    print(f"0: {torch.cuda.memory_allocated()/1e6} MB")
    x0, x1 = tuple_of_tensors
    del tuple_of_tensors
    print(f"1: {torch.cuda.memory_allocated()/1e6} MB")
    z = x0 + 1
    print(f"2: {torch.cuda.memory_allocated()/1e6} MB")
    del x0
    print(f"3: {torch.cuda.memory_allocated()/1e6} MB")
    w = x1 + z + 1
    print(f"4: {torch.cuda.memory_allocated()/1e6} MB")
    del x1, z
    print(f"5: {torch.cuda.memory_allocated()/1e6} MB")
    return w

l = tuple(torch.ones(1000, device="cuda") for _ in range(2))
out = f(l)
# prints
# 0: 0.008192 MB
# 1: 0.008192 MB
# 2: 0.012288 MB
# 3: 0.012288 MB
# 4: 0.016384 MB
# 5: 0.012288 MB
```

How do you think we could support both immutable and mutable containers? Maybe it's better to avoid using `tree_iter` here and use plain iterations over args if args is a tuple?

Reply:
Your point makes a lot of sense. Indeed, when I pass a list as an argument to `backward_fn`, the peak memory consumption decreased.

I suppose this can be simulated by interpreting `clear_mutable_collection` in the same way as `del`. This involves keeping track of how many lists contain which tensors, but this is apparently doable.